python/llm/src/bigdl/llm/transformers/linear_quant.py

```diff
@@ -221,18 +230,45 @@ def forward(self, x: torch.Tensor):
 
         x_shape = x.shape
         x_2d = x.view(-1, x_shape[-1])
+        if x_2d.is_contiguous() is False:
+            x_2d = x_2d.contiguous()
 
         x0 = self.weight.data
 
-        # todo may need to set a different number on different platforms
-        if IS_SERVER and self.qtype == SYM_INT4 and x_2d.shape[0] >= TORCH_LINEAR_THRESHOLD:
-            x0_fp32 = ggml_int4_convert_fp32(x0, self.weight_shape, self.weight_length)
-            result = F.linear(x, x0_fp32, self.bias)
-        else:
-            result = ggml_matmul_src1_x_src0_t(x0, x_2d, self.weight_shape, self.qtype)
+        if x0.device.type == "xpu":
+            # xpu only support q4_0 now
+            invalidInputError(self.qtype == 2,
+                              "XPU int4 only support q4_0(qtype=2) now.")
+            # GPU logic
+            # use just in time compile now
+            # just compile this int4 linear extension at first time
+            try:
+                import linear_q4
+            except ModuleNotFoundError:
+                from torch.xpu.cpp_extension import load
+                linear_q4 = load(name="linear_q4",
+                                 sources=[os.path.join(os.path.dirname(__file__), 'linear_xpu.cpp'),
+                                          os.path.join(os.path.dirname(__file__),
+                                                       'linear_xpu_kernel.cpp')],
+                                 extra_ldflags=["-fsycl"])
```

Comment:
TODO: ahead-of-time compilation instead of JIT (maybe in next PR)

Reply:
ok, have no idea how to support AOT is suitable now, I will  consult ipex team for some suggestions and maybe support it in next PR.